hadoop-hdfs-project/hadoop-hdfs-native-client/src/main/native/libhdfs/jni_helper.c

```diff
@@ -534,6 +545,21 @@ JNIEnv* getJNIEnv(void)
 
     state->env = getGlobalJNIEnv();
     mutexUnlock(&jvmMutex);
+
+    if (!jclassesInitialized) {
+      mutexLock(&jclassInitMutex);
+      if (!jclassesInitialized) {
+        jthrowable jthr = NULL;
+        jthr = initCachedClasses(state->env);
+        if (jthr) {
+          printExceptionAndFree(state->env, jthr, PRINT_EXC_ALL,
+                                "initCachedClasses failed");
+          return NULL;
```

Comment:
does this return (and the one above) miss some cleanup that should happen in the 'goto fail' label? (ugh I hate programming in C...)

Reply:
Yeah, fixed. I think the one above is okay; the cleanup is dependent on `ThreadLocalState` being created.